docs/source/testing.rst

```diff
@@ -0,0 +1,867 @@
+is installed it will print the random seed it used for that session,
+e.g:
+
+.. code-block:: bash
+
+   pytest tests
+   [...]
+   Using --random-order-bucket=module
+   Using --random-order-seed=573663
+
+So that if the given particular sequence fails, you can reproduce it by
+adding that exact seed, e.g.:
+
+.. code-block:: bash
+
+   pytest --random-order-seed=573663
+   [...]
+   Using --random-order-bucket=module
+   Using --random-order-seed=573663
+
+It will only reproduce the exact order if you use the exact same list of
+tests (or no list at all (==all)). Once you start to manually narrowing
+down the list you can no longer rely on the seed, but have to list them
+manually in the exact order they failed and tell pytest to not randomize
+them instead using ``--random-order-bucket=none``, e.g.:
+
+.. code-block:: bash
+
+   pytest --random-order-bucket=none tests/test_a.py tests/test_c.py tests/test_b.py
+
+To disable the shuffling for all tests:
+
+.. code-block:: bash
+
+    pytest --random-order-bucket=none
+
+By default ``--random-order-bucket=module`` is implied, which will
+shuffle the files on the module levels. It can also shuffle on
+``class``, ``package``, ``global`` and ``none`` levels. For the complete
+details please see its `documentation <https://github.com/jbasko/pytest-random-order>`__.
+
+Randomization alternatives:
+
+-  ```pytest-randomly`` <https://github.com/pytest-dev/pytest-randomly>`__
+
+This module has a very similar functionality/interface, but it doesn't
+have the bucket modes available in ``pytest-random-order``. It has the
+same problem of imposing itself once installed.
+
+Look and feel variations
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+pytest-sugar
+^^^^^^^^^^^^
+
+`pytest-sugar <https://github.com/Frozenball/pytest-sugar>`__ is a
+plugin that improves the look-n-feel, adds a progressbar, and show tests
+that fail and the assert instantly. It gets activated automatically upon
+installation.
+
+``pip install pytest-sugar``
+
+To run tests without it, run:
+
+.. code-block:: bash
+
+    pytest -p no:sugar
+
+or uninstall it.
+
+
+
+Report each sub-test name and its progress
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For a single or a group of tests via ``pytest`` (after
+``pip install pytest-pspec``):
+
+.. code-block:: bash
+
+   pytest --pspec tests/test_optimization.py 
+
+
+
+Instantly shows failed tests
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+`pytest-instafail <https://github.com/pytest-dev/pytest-instafail>`__
+shows failures and errors instantly instead of waiting until the end of
+test session.
+
+.. code-block:: bash
+
+    pip install pytest-instafail
+
+.. code-block:: bash
+
+    pytest --instafail
+
+To GPU or not to GPU
+~~~~~~~~~~~~~~~~~~~~
+
+On a GPU-enabled setup, to test in CPU-only mode add
+``CUDA_VISIBLE_DEVICES=""``:
+``CUDA_VISIBLE_DEVICES="" pytest tests/test_vision.py``
+
+XXX: Currently we have ``USE_CUDA=1`` which is currently False by default, but most likely this will change.
```

Comment:
OK, it'd just be tricky to remember that this section needs work, as I'm first waiting for some code changes, but I don't know when this will be approved. I will put it on my todo list.

I'm curious - why is it OK to have XXX in the code, but not in the docs? especially developer-oriented (i.e. end users most likely won't read this). Is it XXX that's the problem and not the fact that there is a note suggesting more to come in the future/this section is incomplete? Would TODO be acceptable?